Library/Homebrew/brew.sh

```diff
@@ -247,7 +247,11 @@ then
   # Hide shellcheck complaint:
   # shellcheck source=/dev/null
   source "$HOMEBREW_BASH_COMMAND"
-  { update-preinstall; "homebrew-$HOMEBREW_COMMAND" "$@"; exit $?; }
```

Comment:
You're right, Bash normally reads line-by-line. But that's not what we're relying on here. Instead it's the curly braces (search for “group command” in the Bash man page) that cause the expression to be read as a whole before it gets executed. Whether it extends over several lines or not doesn't matter.

Here are four self-modifying scripts that demonstrate both your point (normally execution line by line) and mine (the different behavior of a group command):

```sh
#!/bin/bash
self="$(<"$0")"
sed -e 's/H[e]llo/Goodbye/' <<<"$self" >"$0"
echo "test-1: Hello World!"
exit 0
```

```sh
#!/bin/bash
self="$(<"$0")"
{
  sed -e 's/H[e]llo/Goodbye/' <<<"$self" >"$0"
  echo "test-2: Hello World!"
  exit 0
}
```

```sh
#!/bin/bash
self="$(<"$0")"
sed -e 's/H[e]llo/Goodbye/' <<<"$self" >"$0"; echo "test-3: Hello World!"; exit 0
```

```sh
#!/bin/bash
self="$(<"$0")"
{ sed -e 's/H[e]llo/Goodbye/' <<<"$self" >"$0"; echo "test-4: Hello World!"; exit 0; }
```

And here's their output:

```console
$ ./test-1
test-1: Goodbye World!
$ ./test-2
test-2: Hello World!
$ ./test-3
test-3: Hello World!
$ ./test-4
test-4: Hello World!
```